### DISCUSSION TITLE
Make environment files optional

### DISCUSSION THREAD

**Discussion by tom-combet** [phpro/grumphp#993] [Ideas]
As a developer, I want to add optional environment files so that I can allow local environment files to final users without versioning them.

Currently, GrumPHP throws an exception if environment file is not found:
`PHP Fatal error:  Uncaught _HumbugBox84aa00e62dca\Symfony\Component\Dotenv\Exception\PathException: Unable to read the ".env.local" environment file.`

Is it normal behavior or should I create a pull request?

**Comment by veewee** [comment]
I'm a bit confused here.
Normally GrumPHP only loads env files that are configured directly inside grumphp.yaml.

Can you share more details about your setup and your configuration file?

**Reply by tom-combet** [reply]
GrumPHP loads indeed correct env files I defined in the configuration:

```yaml
environment:
  files:
    - .env
    - .env.local
```

There is no bug there. However, I'd expect that GrumPHP doesn't try to load `.env.local` file if it's not present.
In fact, I don't version `.env.local` so final user may not have it when installing my project. He can add it later, depending on his needs.

**Reply by veewee** [reply]
If you look to it from the opposite way: isn't it strange that you define a file to load in the config file and it doesn't give you an exception if it is not there?

It also works like this in docker-compose e.g.

Are there specific env vars you are trying to make configurable and are there other options to do so?

**Reply by tom-combet** [reply]
My particular use case is to allow user to define his php executable path.
Local environment file is probably the more straightforward way to do it, and I don't really know if I will be able to find a workaround.

Feel free to check the [grumphp.yml](https://github.com/Kaudaj/prestashop-maker/blob/main/tests/php/grumphp.dist.yml) file for more informations.

**Reply by tom-combet** [reply]
Hi @veewee, could you reconsider the case please?
I still didn't find any way to version default values while allowing non-versioned custom ones at user level...

**Reply by veewee** [reply]
It still doesn't make sense to me to specify a file that doesn't exists in grumphp.yaml (similar to docker-compose).

Your case is still the PHP path? 
Why do you want to configure that path?

By default, grumphp detects the PHP version based on your PATH - so you could prepend the path as well?
There are plenty of tools to switch the default php binary as well.
For example: if your project runs on PHP 7.4, then it should use php7.4 for that project anyways. What would the benifit be of overwriting that in .env.local instead of making sure the php command loads the php7.4 binary?

**Reply by tom-combet** [reply]
Ah, makes sense for the php path, yup my bad,
About .env files, for me it makes sense to use an optional local one ([example](https://harrisonpim.com/blog/you-should-commit-your-env-files-to-version-control)). But anyway, you solved my use case for now, thanks!

**Reply by tom-combet** [reply]
And is it possible when you have to check for multiple versions of PHP?

**Reply by veewee** [reply]
Sure is possible: you use CI for that

example:

https://github.com/phpro/grumphp/blob/master/.github/workflows/grumphp.yml#L10-L15

**Reply by tom-combet** [reply]
I meant locally, with GrumPHP. I already run CI in different php versions.

**Reply by veewee** [reply]
Sure, if you prefix the path to the  exact PHP executable you want to use.

```
PATH="/xxx/php8.1:$PATH" grumphp run
```

That way you can quickly test multiple versions.

**Reply by tom-combet** [reply]
Yup, but I can't automate the process for the user then, cause I can't predict his executable path.
Furthermore, I'd like to only run php against the files, not whole grum process. So it concerns only some tasks (see [there](https://github.com/Kaudaj/kjmodulebedrock/blob/dev/tests/grumphp.dist.yml)).

**Reply by veewee** [reply]
Why would you expect people working on your package to have 4-6 php versions configured locally?
That alone is a tremendous getting started to contribute step.

Since all it does is php -l, wouldn't it make sense to either do it on CI, or alternatively through docker.

To me it seems more like a CI job instead of a code quality check that you quickly run before committing nevertheless.

**Reply by tom-combet** [reply]
> Why would you expect people working on your package to have 4-6 php versions configured locally?
> That alone is a tremendous getting started to contribute step.

It wouldn't be compulsory, but for those who wants to check locally it could be nice.

> To me it seems more like a CI job instead of a code quality check that you quickly run before committing nevertheless.

Well also true, I'll keep it for CI only I guess. Thanks again.
